h2o-core/src/main/java/water/parser/CharSkippingBufferedString.java

```diff
@@ -0,0 +1,114 @@
+package water.parser;
+
+import water.MemoryManager;
+
+import java.util.Arrays;
+
+/**
+ * A schema over a string represented as an array of bytes.
+ * This schema enables characters to be skipped inside the string. unlike the basic {@link BufferedString}
+ * Skipped characters are not serialized by toString method.
+ */
+class CharSkippingBufferedString {
+
+    private int[] _skipped;
+    private int _skippedWriteIndex;
+    private BufferedString _bufferedString;
+
+    CharSkippingBufferedString() {
+        _skipped = new int[0];
+        _skippedWriteIndex = 0;
+        _bufferedString = new BufferedString();
+    }
+
+    protected void addChar() {
+        _bufferedString.addChar();
+    }
+
+    protected void removeChar() {
+        _bufferedString.removeChar();
+    }
+
+    protected byte[] getBuffer() {
```

Comment:
Why do we need this one? It doesn't consider the skipped characters and it could be miss-used?

Reply:
## Length + Offset
:heavy_check_mark: We can get rid of those.

This was used to calculate chunk overflow. 
```
          if((str.getOffset() + str.length()) > str.getBuffer().length){ // crossing chunk boundary
```

The very fact this formula does not subtract skipped characters is correct. In this place of CsvParser, we need to know whether we've reached **reading** from the original chunks representing the content of the file. Replaced with a single, better named method.

## getBuffer
:no_entry: 

This one is used in CSV parser in an **assert** on line 147:

```java
          if(str.isOverflown()){ // crossing chunk boundary
            assert str.getBuffer() != bits;
            str.addBuff(bits);
          }
```

Reply:
original is fine because nothing is done to the buffer ;)

it is not a big deal but it is also no the most developer friendly